### PR TITLE
search: alert for possible structural search pattern

### LINKED ISSUE #11593: Add structural search help message
- **Sourcegraph version: All versions
- **Platform information:** cloud
#### Steps to reproduce:

1. Search useState(:[arg]) or any structural search. 
2. Forget to click the structural search button.

#### Expected behavior:
See a help message under 'No search results' message that reads:

 "It looks like you are using structural search . You must enable structural search by clicking the [ ] button to use those language features."  

Alternatively, (depending on complexity and how accurate we could be), detect structural search in a query, and automatically apply the button. 

#### Actual behavior:

I see no search results and wonder if my syntax is wrong.

### PR DESCRIPTION
**Pull request by rvantonder**
Resolves #11593. The alert is raised if:

- There are no results AND
- The current search mode is not structural search AND
- The query contains some hole syntax like `:[...]`

It looks like this:

Screen Shot 2020-06-24 at 11 31 59 PM

Tested with integration test.

### CODE REVIEW

**Comment by rvantonder** [comment]
@rrhyne while I have your attention :-) This is a nice representative alert of what the current notifications look like and I'm confident we can improve the design and experience here, maybe it will stimulate your thinking. Also I'm not super attached to the wording, so feel free to suggest otherwise.

File: cmd/frontend/graphqlbackend/search_results.go [hunk 1]
```diff
@@ -1772,6 +1772,14 @@ func (r *searchResolver) doResults(ctx context.Context, forceOnlyResultType stri
 		alert = newAlert // takes higher precedence
 	}
 
+	if len(results) == 0 && r.patternType != query.SearchTypeStructural {
+		log15.Info("ok", "", args.PatternInfo.Pattern)
+		if matchHoleRegexp.MatchString(r.originalQuery) {
+			log15.Info("set")
```
Comment (asdine) [inline cmd/frontend/graphqlbackend/search_results.go:1778]: Are these leftover debug logs?

Comment (rvantonder) [reply cmd/frontend/graphqlbackend/search_results.go]: Yup! thanks :)

**Comment by rrhyne** [comment]
This is perfect!

**Comment by rvantonder** [comment]
@rrhyne asks:

> Just curious, how well can we detect structural search intensions? Could we detect it and enable it without warning Possibly a with message like google does. It spellchecks, shows you results for the proper spelling but lets you search for the incorrect spelling as well.

This touches on what I mention in https://github.com/sourcegraph/sourcegraph/issues/9745#issuecomment-613746593 which is along the lines of "we didn't find anything but here are results for what we think you could have meant", and I think we want this general capability.

For your specific comment, there are different confidence levels about inferring what users could have meant. You mentioned one case here for structural search, and our measure basically depends on detecting the particular hole syntax. I would give that "moderate" confidence, but not high confidence. What I'd like for search is a general mechanism to communicate these if we wanted to (regardless of what confidence level we have), which we could then use to gauge utility Like in that structural search case, maybe it works well, maybe it doesn't, but we don't really have a good way to test/deploy this because I feel like our alerts need to first better communicate any proactive decision we are taking on user input, and there isn't a good way to do that (it is currently impossible to show both an alert/notification and results). So overall I think our existing alert/suggest mechanism is insufficient to communicate this general mechanism in the flavor of Google-style "we interpreted your query as X, did you mean X or Y?" which we need before implementing the behavior you suggest. Eager to see better UX/UI designs for exactly this use case (again see, #9745).

**Pull request merged**